docs_src/design/ucr/Microservice-Authentication.md

```diff
@@ -0,0 +1,146 @@
+  access that was previously granted,
+  or allow customers to tie in to enterprise
+  identity management systems.
+
+For purposes of this UCR, microservice authentication implies
+that the receiving microservice has access to the identity
+of the caller and can write program logic based on that identity.
+
+
+### Existing solutions
+
+Microservice authentication is currently implemented around two primary vectors:
+
+- Token-based authentication schemes.
+  Initiator sends an identifier along with a request to the receiver.
+  The identifier is cryptographically validated using a key trusted by
+  the receiver, or the receiver asks a trusted third party to verify the identifier.
+
+  A benefit of token-based authentication schemes is identity delegation,
+  whereby the identifier can be passed through a chain of calls to
+  preserve the identity of the original initiator.
+  The identifier can often be tunned through other protocols.
+  Another benefit of token-based authentication is that
+  it flows easily through a web application firewall.
+
+  A drawback of token-based authentication is that due to man-in-the-middle
+  threats, token-based authentication over an unencrypted network is insecure.
+  Another drawback of token-based authentication is that it is unidirectional:
+  the receiver can authenticate the initiator, but not vice-versa.
+
+- Mutual-auth TLS.
+  Both the initiator and the receiver participate in a session-oriented message
+  exchange based on public-key cryptography, where each trusts the other's
+  digital signature authority.
+
+  A benefit of mutual-auth TLS is that it enforces proof-of-possession
+  (of a cryptographic key), thus preventing token-stealing attacks.
+
+  A drawback of mutual-auth TLS is that it requires layer 4
+  (IP:port) network connectivity.
+  Layer 7 (e.g. HTTP) interpretation of the traffic stream,
+  such as a URL-rewriting HTTP router, breaks end-to-end encryption
+  and the ability to cryptographically identify the caller.
+  Another drawback is that mutual-auth TLS requires encryption
+  for all use cases, including same-host communication.
+  Encryption may be redundant if encryption is implemented
+  at lower layers in the network stack.
+  Combining authentication and encryption may also complicate debugging.
+
+  The most onerous drawback to mutual-auth TLS is overhead associated with
+  certificate and key rotation on both the initiator and receiver when
+  the cryptoperiods of their cryptographic materials expire.
+
+  Many service mesh products are based on mutual-auth TLS
+  for confidentiality and enforcement of network policy,
+  but application access to the authenticated caller identity is not a given,
+  and varies from product to product.
+
+
+### Requirements
+
+While running in secure mode, each EdgeX service must
+receive requests over an authenticated interface
+and must provide authentication information
+to components that require it.
```

Comment:
Is providing "authentication information" for informing unauthenticated clients how to get authenticated? Or is it for enabling bi-directional authentication?

Reply:
Thanks for pointing this out.  Will rewrite this as two separate requirements.  The first is, when calling an EdgeX microservice, if secure mode is enabled, make sure that the request is credentialed.   The second is, when receiving a request, to require a valid credential when secure mode is disabled.    If enabling security on a microservice-level, this means that I security enabled microservice can call one with security disabled, but not the other way around.